python/src/spark_rapids_ml/umap.py

```diff
@@ -0,0 +1,426 @@
+        super().__init__()
+        self.set_params(**kwargs)
+        self.sample_fraction = sample_fraction
+
+    def _create_pyspark_model(self, result: Row) -> "UMAPModel":
+        return UMAPModel.from_row(result)
+
+    def _fit(self, dataset: DataFrame) -> "UMAPModel":
+        if self.sample_fraction < 1.0:
+            data_subset = dataset.sample(
+                withReplacement=False,
+                fraction=self.sample_fraction,
+                seed=self.cuml_params["random_state"],
+            )
+        else:
+            data_subset = dataset
+
+        input_num_workers = self.num_workers
+        # Force to single partition, single worker
+        self._num_workers = 1
+        if data_subset.rdd.getNumPartitions() != 1:
+            data_subset = data_subset.repartition(1)
+
+        pipelined_rdd = self._call_cuml_fit_func(
+            dataset=data_subset,
+            partially_collect=False,
+            paramMaps=None,
+        )
+        rows = pipelined_rdd.collect()
+        model = self._create_pyspark_model(rows[0])
+        model._num_workers = input_num_workers
+
+        self._copyValues(model)
+        self._copy_cuml_params(model)  # type: ignore
+
+        return model
+
+    def _fit_array_order(self) -> _ArrayOrder:
+        return "C"
+
+    def _get_cuml_fit_func(
+        self,
+        dataset: DataFrame,
+        extra_params: Optional[List[Dict[str, Any]]] = None,
+    ) -> Callable[[FitInputType, Dict[str, Any]], Dict[str, Any],]:
+        cls = self.__class__
+        array_order = self._fit_array_order()
+
+        def _cuml_fit(
+            dfs: FitInputType,
+            params: Dict[str, Any],
+        ) -> Dict[str, Any]:
+            from cuml.manifold import UMAP as CumlUMAP
+
+            umap_object = CumlUMAP(
+                **params[param_alias.cuml_init],
+            )
+
+            df_list = [x for (x, _, _) in dfs]
+            if isinstance(df_list[0], pd.DataFrame):
+                concated = pd.concat(df_list)
+            else:
+                concated = _concat_and_free(df_list, order=array_order)
+
+            if dfs[0][1] is not None:
+                # If labels are provided, call supervised fit
+                label_list = [x for (_, x, _) in dfs]
+                if isinstance(label_list[0], pd.DataFrame):
+                    labels = pd.concat(label_list)
+                else:
+                    labels = _concat_and_free(label_list, order=array_order)
+                local_model = umap_object.fit(concated, y=labels)
+            else:
+                # Call unsupervised fit
+                local_model = umap_object.fit(concated)
+
+            dtype = str(local_model.embedding_.dtype)
+
+            return {
+                "embedding_": [local_model.embedding_.tolist()],
+                "raw_data_": [concated.tolist()],
+                "n_cols": len(local_model.embedding_[0]),
+                "dtype": dtype,
```

Comment:
We could consider return the embeddings and raw data in smaller chunks (e.g. based on arrow batch size parameter) with one embedding/data row per row of returned df.   Not sure if having corresponding embedding and data rows in the same pdf row (i.e. two columns) is better than first all embedding rows and then all data rows say, in a single column.   Former seems logically cleaner.

Can n_cols and dtype be determined on driver?